designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,531 @@
+- Start Date: 2019-09-28
+- RFC PR: https://github.com/eslint/rfcs/pull/40
+- Authors: Toru Nagashima ([@mysticatea](https://github.com/mysticatea))
+
+# `ESLint` Class Replacing `CLIEngine`
+
+## Summary
+
+This RFC adds a new class `ESLint` that provides asynchronous API and deprecates `CLIEngine`.
+
+## Motivation
+
+- We have functionality that cannot be supported with the current synchronous API. For example, ESLint verifying files in parallel, formatters printing progress state, formatters printing results in streams etc. A move to an asynchronous API would be beneficial and a new `ESLint` class can be created with an async API in mind from the start.
```

Comment:
> The naming of CLIEngine was to make it clear that this class was to be used for creating CLI programs and not for general purpose linting

I have an objection around this.

In the real world, also GUIs such as editor integrations use `CLIEngine` because the `Linter` class lacks important functionality the users expected (e.g., detect config files, parse config files, load parsers/configs/plugins, ...). That means people have understood our API as "the `CLIEngine` class is the primary class and the `Linter` class is the special extracted part for browsers." Therefore, I think that the ideal names are:

- the current `CLIEngine` => `ESLint` or `Linter`
- the current `Linter` => `LinterForBrowsers`, `LintEngine`, `LinterCore`, `LinterKernel`, or something like.

![image](https://user-images.githubusercontent.com/1937871/68723299-8258a880-05fb-11ea-889c-b6bfa31ffaee.png)

The problem here is that the class that people can use only a part of ESLint spec has the most general name in our API, and the class that people can use the full of ESLint spec has a too limited name. Therefore, people have tried the `Linter` class first, but it didn't work as expected, then we have seen questions about `Linter` and have said "please use `CLIEngine` instead" many times.

Reply:
I understand your reasoning. The problem is that `Linter` was designed to be the primary object that everyone uses everywhere (including browsers and Node.js) so people are doing the right thing. However, most of the time people looking at the API are wanting to mimic the ESLint CLI in some way, and that's why we have `CLIEngine`. As far as I know, the most common cases are 1) tooling plugins (editors, build tools, etc.), and 2) standalone CLI wrappers (Standard, XO). Is there another case you're thinking of?

My problem with calling this `ESLint` is that it is completely limited to running in Node.js, and I think people trying to use the API in a browser would be confused as to why `ESLint` isn't available. I just don't think a class that is tied to a specific runtime should be given the name `ESLint`. If you're opposed to including "CLI", that's fine, I just think the class name needs to be explicit that it is not a general-purpose utility and is limited to one runtime or use case.

(A lot of this is because we are dealing with a codebase that is over six years old and the project has grown in ways I didn't anticipate when initially setting up the structure.)

Reply:
Hmm. I want to avoid that the class that doesn't support the important parts that the [Configuring ESLint](https://eslint.org/docs/user-guide/configuring) page describes is exposed as the most general name. May you accept to rename `Linter` as well? In my head, `ESLint` => `LinterShell` and `Linter` => `LinterKernel`.

Reply:
I have remembered that [we don't support ESLint in browsers officially](https://github.com/eslint/eslint/issues/8348). Doesn't your concern conflict with our stance?

Reply:
I don't think we should rename `Linter` as part of this. That's just introducing more breaking changes without much benefit.

> I have remembered that we don't support ESLint in browsers officially. Doesn't your concern conflict with our stance?

No. I think it's reasonable to assume that a class named `ESLint` should work in a browser, just like it's reasonable to assume that `Linter` would work in the browser (even if we don't formally support it, people still use it). Again, I purposely named the class `CLIEngine` to make it clear that this was to be used for creating CLI applications. I think whatever replaces it should be named equally clearly.

I can live with `LinterShell`.

Reply:
The priority order I thought is different from yours.

1. End-users expect that they can configure ESLint by the way the "[Configuring ESLint](https://eslint.org/docs/user-guide/configuring)" page describes because it's the only page of the official document about configuration. Therefore, API users expect that the API of ESLint supports it.
1. API users expect that the API works in browsers even if we don't support it officially.

To me, (1) is pretty more important than (2). Therefore, I think that the API that supports full-spec should have the most general name, and the `Linter` should a specific-name that means workable without file systems.

---

Anyway, `LinterShell` looks a compromise for us.